File: typescript/src/model/statements/wdl-import.ts
/** Import statement nodes for the TypeScript WDL model. */
import type { WdlStringLiteral } from '../expressions/wdl-string-literal.js';

/** Base class for the three WDL import forms. */
export abstract class WdlImport {
  private importIdentifierValue: string | undefined;
  private sourceTextValue: string | undefined;

  /** Creates an import from its source literal. */
  public constructor(protected sourceValue?: WdlStringLiteral) {}
  /** Returns the source literal naming the imported document. */
  public getSource(): WdlStringLiteral | undefined {
    return this.sourceValue;
  }
  /** Sets the source literal naming the imported document. */
  public setSource(source: WdlStringLiteral | undefined): void {
    this.sourceValue = source;
  }

  /** Returns the resolved identifier used to index imported documents. */
  public getImportIdentifier(): string | undefined {
    return this.importIdentifierValue;
  }

  /** Sets the resolved identifier used to index imported documents. */
  public setImportIdentifier(importIdentifier: string | undefined): void {
    this.importIdentifierValue = importIdentifier;
  }

  /** Returns the raw source text loaded for this import, when available. */
  public getSourceText(): string | undefined {
    return this.sourceTextValue;
  }

  /** Sets the raw source text loaded for this import. */
  public setSourceText(sourceText: string | undefined): void {
    this.sourceTextValue = sourceText;
  }
}

/** Models a single imported member and optional local alias. */
export class WdlImportMember {
  /** Creates an import member from its source name and optional local alias. */
  public constructor(
    private memberValue?: string,
    private aliasValue?: string,
  ) {}
  /** Returns the imported member name. */
  public getMember(): string | undefined {
    return this.memberValue;
  }
  /** Sets the imported member name. */
  public setMember(member: string | undefined): void {
    this.memberValue = member;
  }
  /** Returns the local alias for the imported member. */
  public getAlias(): string | undefined {
    return this.aliasValue;
  }
  /** Sets the local alias for the imported member. */
  public setAlias(alias: string | undefined): void {
    this.aliasValue = alias;
  }
}

/** Models the standard `import "x" as y` form with optional alias clauses. */
export class WdlImportStandard extends WdlImport {
  private readonly memberValues: WdlImportMember[] = [];
  private aliasValue: string | undefined;
  /** Returns the import namespace alias. */
  public getAlias(): string | undefined {
    return this.aliasValue;
  }
  /** Sets the import namespace alias. */
  public setAlias(alias: string | undefined): void {
    this.aliasValue = alias;
  }
  /** Returns the ordered imported type/member alias clauses attached to the import. */
  public members(): WdlImportMember[] {
    return this.memberValues;
  }
}

/** Models `import * from "x"`. */
export class WdlImportStar extends WdlImport {}

/** Models `import { a, b as c } from "x"`. */
export class WdlImportMembers extends WdlImport {
  private readonly memberValues: WdlImportMember[] = [];
  /** Returns the ordered imported members selected by this import. */
  public members(): WdlImportMember[] {
    return this.memberValues;
  }
}

File: python/src/wdl_model/model/statements/wdl_import.py
"""Import statement nodes for the Python WDL object model.

WDL supports standard imports, star imports, and member-list imports. See the specification section
"Import Statements".
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from wdl_model.model.base import WdlSourceRange
from wdl_model.model.expressions import WdlStringLiteral


@dataclass
class WdlImport:
    """Base import statement carrying the imported source literal."""

    source: WdlStringLiteral | None = None
    importIdentifier: str | None = None
    sourceText: str | None = None
    source_range: WdlSourceRange | None = None


@dataclass
class WdlImportMember:
    """Single imported member with an optional local alias."""

    member: str | None = None
    alias: str | None = None


@dataclass
class WdlImportStandard(WdlImport):
    """Standard import form: import source [as alias] with optional type alias clauses."""

    alias: str | None = None
    _members: deque[WdlImportMember] = field(default_factory=deque)

    def members(self) -> deque[WdlImportMember]:
        """Return the ordered imported type alias clauses attached to the import."""
        return self._members


@dataclass
class WdlImportMembers(WdlImport):
    """Member-list import form: import { a, b as c } from source."""

    _members: deque[WdlImportMember] = field(default_factory=deque)

    def members(self) -> deque[WdlImportMember]:
        """Return the ordered imported members selected by the import."""
        return self._members


@dataclass
class WdlImportStar(WdlImport):
    """Star import form: import * from source."""

    pass
